File: app/src/main/java/com/worldventures/dreamtrips/wallet/service/command/firmware/LoadPuckAtmelFirmwareCommand.java
package com.worldventures.dreamtrips.wallet.service.command.firmware;

import io.techery.janet.command.annotations.CommandAction;
import io.techery.janet.smartcard.action.support.UpgradeExtAtmelFirmwareAction;
import io.techery.janet.smartcard.event.UpgradeExtAtmelFirmwareProgressEvent;
import rx.Observable;

import com.worldventures.dreamtrips.wallet.domain.entity.ImmutableSmartCardFirmware;
import com.worldventures.dreamtrips.wallet.domain.entity.SmartCardFirmware;

import java.io.File;

@CommandAction
public class LoadPuckAtmelFirmwareCommand extends BaseLoadFirmwareCommand {

   private final File firmwareFile;
   private final String firmwareVersion;

   public LoadPuckAtmelFirmwareCommand(File firmwareFile, String firmwareVersion) {
      this.firmwareFile = firmwareFile;
      this.firmwareVersion = firmwareVersion;
   }

   @Override
   Observable<Integer> provideProgress() {
      return janet.createPipe(UpgradeExtAtmelFirmwareProgressEvent.class)
            .observeSuccess()
            .map(event -> event.progress);
   }

   @Override
   Observable<Void> loadFile() {
      return janet.createPipe(UpgradeExtAtmelFirmwareAction.class)
            .createObservableResult(new UpgradeExtAtmelFirmwareAction(firmwareFile))
            .map(action -> (Void) null);
   }

   @Override
   SmartCardFirmware updatedSmartCardFirmware(SmartCardFirmware currentSmartCardFirmware) {
      return ImmutableSmartCardFirmware.copyOf(currentSmartCardFirmware).withExternalAtmelVersion(firmwareVersion);
   }
}

File: wallet/src/main/java/com/worldventures/wallet/service/firmware/command/LoadPuckAtmelFirmwareCommand.kt
package com.worldventures.wallet.service.firmware.command

import com.worldventures.wallet.domain.entity.SmartCardFirmware
import io.techery.janet.command.annotations.CommandAction
import io.techery.janet.smartcard.action.support.UpgradeExtAtmelFirmwareAction
import io.techery.janet.smartcard.event.UpgradeExtAtmelFirmwareProgressEvent
import rx.Observable
import java.io.File

@CommandAction
class LoadPuckAtmelFirmwareCommand(private val firmwareFile: File, private val firmwareVersion: String) : BaseLoadFirmwareCommand() {

   override fun provideProgress(): Observable<Int> {
      return janet.createPipe(UpgradeExtAtmelFirmwareProgressEvent::class.java)
            .observeSuccess()
            .map { event -> event.progress }
   }

   override fun loadFile(): Observable<Void> {
      return janet.createPipe(UpgradeExtAtmelFirmwareAction::class.java)
            .createObservableResult(UpgradeExtAtmelFirmwareAction(firmwareFile))
            .map { null }
   }

   override fun updatedSmartCardFirmware(currentSmartCardFirmware: SmartCardFirmware) =
         currentSmartCardFirmware.copy(externalAtmelVersion = firmwareVersion)
}
